### PR TITLE
NF: Add UniqueArrayList data structure

### PR DESCRIPTION
**Pull request by TarekkMA**
## Purpose / Description

A base PR for #8322

## Approach

`UniqueArrayList` is a `List` compliant data structure, that uses an internal `TreeSet` with optional comparator, with `ArrayList`.

### My thought process when I created  `UniqueArrayList` 

The `UniqueArrayList<E>` class is created to be able to do the following:
*  **Maintain the order of the items**
*  **Indexable**: getting index of an element and vice versa
   * Needed in recycler view where it requires access to element via index a lot while scrolling.
*  **Sortable**: need a way to sort the array
    * Needed when adding tags or loading the Dialog, where checked tags must be shown first
*   **Unique**: maintaining a unique list of items using comparators
    * making tags unique with `String.CASE_INSENSITIVE_ORDER`
    
👎  `ArrayList<E>`: not Unique
👎   `TreeSet<E>`: doesn't maintain the order and no indexing and not sortable
👎  `LinkedHashSet<E>`: not indexable and cannot be sorted and not uniqe with custom comparator
👎  [`SetUniqueList<E>`](https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/list/SetUniqueList.html): almost perfect but not sortable and no comparator support.

## How Has This Been Tested?

Unit tests

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+    public static <E> UniqueArrayList<E> fromCollection(@NonNull Collection<? extends E> c, @Nullable Comparator<? super E> comparator) {
+        Set<E> set;
+        if (comparator != null) {
+            set = new TreeSet<>(comparator);
+        } else {
+            set = new TreeSet<>();
+        }
+        final List<E> l2 = new ArrayList<>();
+        for (E e : c) {
+            if (!set.contains(e)) {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:87]:
```java
bool new_string = set.add(e);
if (new_string) {
  l2.add(e);
}
```
Usually, it's better to add directly, since it returns the value of `contains` https://docs.oracle.com/javase/7/docs/api/java/util/Set.html#add(E)

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+     * any subsequent elements to the right (adds one to their indices).
+     *
+     * @param index index at which the specified element is to be inserted
+     * @param element element to be inserted
+     * @deprecated use {@link #addAt(int, Object)} instead
+     */
+    @Override
+    @Deprecated
+    public void add(int index, E element) {
+        if (contains(element)) {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:134]:
```java
bool newString = mSet.add(element);
if (newString) {
  mList.add(index, element);
}
```

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+
+
+    /**
+     * Appends the specified element to the end of this list if it is unique.
+     *
+     * @param e element to be appended to this list
+     * @return true if the element is added
+     */
+    @Override
+    public boolean add(E e) {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:113]: Same remark as add below

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+    /**
+     * Inserts the specified element at the specified position in this
+     * list if it is unique. Shifts the element currently at that position (if any) and
+     * any subsequent elements to the right (adds one to their indices).
+     *
+     * @param index index at which the specified element is to be inserted
+     * @param element element to be inserted
+     * @return true if the element is added
+     */
+    public boolean addAt(int index, E element) {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:151]: Same remark as other add

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+     *
+     * @param index index of the element to replace
+     * @param element element to be stored at the specified position
+     * @return if replaced, the element previously at the specified position
+     * if element is not unique null.
+     * @see #setAt(int, E)
+     */
+    @Override
+    @Nullable
+    public E set(int index, E element) {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:172]: You should remove the previous element from the set too. Otherwise you can't add it anymore

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+
+    /**
+     * Replaces the element at the specified position in this list with
+     * the specified element if the specified is unique.
+     *
+     * @param index index of the element to replace
+     * @param element element to be stored at the specified position
+     * @return if replaced, the element previously at the specified position
+     * if element is not unique null.
+     * @see #setAt(int, E)
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:168]: Is there a difference between set and setAt ? I can't figure it out

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+     * @return true if this collection changed as a result of the call
+     */
+    @Override
+    public boolean addAll(final int index, @NonNull Collection<? extends E> c) {
+        int cIndex = index;
+        for (E e : c) {
+            if (contains(e)) {
+                continue;
+            }
+            mList.add(cIndex, e);
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:227]: I'd suggest doing `add(cIndex, e)`
In this case, no need to use `addAll`

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+/**
+ * A unique collection of items, fast random access, lookup, maintaining order, and sorting.
+ * <p>
+ * The implementation of this class uses {@link ArrayList} and {@link TreeSet}.
+ */
+public class UniqueArrayList<E> implements List<E> {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:40]: You may also want to add that it implements Set

**Review by Arthur-Milchior — COMMENTED** [review]
Thanks a lot for those unit test. Really nice to see that much of them.

A few nit and I think one real problem. Can you please add regression test for :
* list with "a" at position 1
* replace put "b" at position 1
* add "a"
* 
and check that the list is `{"b", "a"}`

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+    /**
+     * Constructs a unique list containing the elements of the specified
+     * collection, in the order they are returned by the collection's
+     * iterator.
+     *
+     * @param c collection with no duplicates whose elements are to be placed into this list
+     * @param s a set of the same items in c
+     * @throws NullPointerException if the specified collection is null
+     */
+    protected UniqueArrayList(@NonNull List<E> c, @NonNull Set<E> s) {
```
[the remark this answers is not in the document]
Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:65]: Oh, Yeah it's from the previous implementation, where I extended `ArrayList`, I think now all factory methods can be constructors.

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java (hunk 5 above)
Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:172]: yes that is a real issue, I should add a test for it.

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java (hunk 7 above)
Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:227]: Yes, but will use the modified version `addAt` to get back the status.

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java (hunk 6 above)
Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:168]: The same as `add` and `addAt`, `set` will return the replaced element, while `setAt` will return a boolean to indicate if the element is unique and `mSet` did add it as a new element.

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java (hunk 8 above)
Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:40]: There are conflicting default methods on the interfaces, tbh I don't know what a spliterator is.

Min API 24
**Set**
```java
    @Override
    default Spliterator<E> spliterator() {
        return Spliterators.spliterator(this, Spliterator.DISTINCT);
    }
```
**List**
```java
    @Override
    default Spliterator<E> spliterator() {
        return Spliterators.spliterator(this, Spliterator.ORDERED);
    }
```

I can throw 
```java
        throw new UnsupportedOperationException(getClass().getSimpleName() + " doesn't support spliterator");
```

**Comment by TarekkMA** [comment]
Added these tests + fixed issues you pointed 
* `test_set_will_remove_replaced_item`
* `test_set_at_will_remove_replaced_item`
* `test_addAll_withIndex_last_position`

**Review by Arthur-Milchior — COMMENTED** [review]
I'd like you to add "NF:" in front of the commit please. From the user point of view, it do no change anything. Actually, it won't even be exported to the app since the code is not used

**Comment by TarekkMA** [comment]
Oh, so NF means no change to the current functionality of the application, no bug fixes or features added ?

**Comment by Arthur-Milchior** [comment]
Right. Literally it means "Non functional"

**Comment by Arthur-Milchior** [comment]
I'm confused about your used of the word "unique". As far as I understand English (Not my mother tong) an element are always "unique" by itself. Elements can be distinct from elements in the list for example. Comments are slightly confusing here.

I added a commit with various suggestion. At this point, found it easier to write than to explain what to write

**Comment by TarekkMA** [comment]
> Right. Literally it means "Non functional"

Okay, I thought it means no functionality added

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java (hunk 8 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java]: Never mind then. Fine not to add it

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java [hunk 10]
```diff
@@ -0,0 +1,419 @@
...
+    public boolean removeAll(@NonNull Collection<?> c) {
+        mList.removeAll(c);
+        return mSet.removeAll(c);
+    }
+
+
+    @Override
+    public boolean retainAll(@NonNull Collection<?> c) {
+        mList.retainAll(c);
+        return mSet.removeAll(c);
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:361]: `retainAll`

**Review by Arthur-Milchior — COMMENTED** [review]
I'm sorry, but I really do not understand when you use the comparator. All iterations are done on the list, I see no place where you call any sorting method. There really seems to be something missing here

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java (hunk 10 above)
Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:361]: Oh I should add tests for these methods I added from the List interface.

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java [hunk 11]
```diff
@@ -178,33 +190,44 @@ public boolean addAt(int index, E element) {
...
         checkIndex(index);
         if (mSet.add(element)) {
             E old = mList.set(index, element);
             mSet.remove(old);
-            return old;
+            return Optional.ofNullable(old);
         }
-        return null;
+        return Optional.empty();
+    }
```
Comment (TarekkMA) [inline AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java]: I do like the `Optional`, but it's an API24 feature.

https://developer.android.com/reference/java/util/Optional

I can refactor, `setAt` to reuse `set` code

**Comment by TarekkMA** [comment]
> I'm sorry, but I really do not understand when you use the comparator. All iterations are done on the list, I see no place where you call any sorting method. There really seems to be something missing here

If you mean the comparator that I pass to `UniqueArrayList`, it used to check equality not used at all for sorting.

Sorting is done via `Collection.sort(comparator)` or `List#sort(comparator)`(API24), but talking about this remind me that the `ListIterator` used for sorting is kinda dangerous, because it has the ability to change the underlying list, thus making it out of sync with the set. 

But sorting methods don't remove items without adding them back.

This high level implementation of sorting found in the docs.
https://developer.android.com/reference/java/util/List#sort(java.util.Comparator%3C?%20super%20E%3E) 

Maybe we should subclass `ListIterator` to add some guards when using it.

**Comment by TarekkMA** [comment]
Will add more tests for untested methods, make the docs clearer.

Finally, look into the `ListIterator` issue. (all the ways the collection can be modified should be intercepted)

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java (hunk 11 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java]: Oh, I didn't realize.
I'd appreciate to avoid code duplication if possible, that would be great. Other solutions are fine too

**Comment by Arthur-Milchior** [comment]
There is an easier solution. You can stop implementing List and just implement the method you expect to need. Keeping compatibility is nice but not mandatory. This will ensure that uses cases that you didn't expect won't be used

**Comment by TarekkMA** [comment]
> There is an easier solution. You can stop implementing List and just implement the method you expect to need. Keeping compatibility is nice but not mandatory. This will ensure that uses cases that you didn't expect won't be used

Issue is I don't know if from 1 year from now it will be misused, I can throw an exception when trying to get a `ListIterator` and do call the sorting methods internally instead of going through outside APIs for sorting

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java [hunk 12]
```diff
@@ -227,6 +250,10 @@ public boolean addAll(@NonNull Collection<? extends E> c) {
     /**
      * Adds all unique elements to the list at a specified index
      *
+     * It is important to note that the uniqueness is checked at addition time, in the order of the elements in `c`.
+     * That means that, [1, 2, 3, 4, 5].addAll(0, [5, 4, 3, 2, 1, 0]) will result in [0, 2, 3, 4, 5] because all elements before 0 were found to be present in the list
+     * However [1, 2, 3, 4, 5].addAll(0, [0, 1, 2, 3, 4, 5]) will result in [0, 1, 2, 3, 4, 5]
+     *
```
Comment (TarekkMA) [inline AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:256]: I don't know what do you mean by this comment, I have added these 2 tests and the result is the same.

`addAt/add(index, element)` they will make way (go right) for the new element and not replace any old one.  
https://docs.oracle.com/javase/7/docs/api/java/util/List.html#add(int,%20E)

* Also `addAt/add(index, element)` index can be `list.size()` making them last element
    * I have added a test to make sure that will work

```java

    @Test
    public void test_addAll_order_1() {
        UniqueArrayList<Integer> uniqueArrayList = new UniqueArrayList<>();

        final List<Integer> l1 = Arrays.asList(1, 2, 3, 4, 5);
        assertTrue(uniqueArrayList.addAll(l1));
        assertSameLists(l1, uniqueArrayList);

        final List<Integer> l2 = Arrays.asList(5, 4, 3, 2, 1, 0);
        assertTrue(uniqueArrayList.addAll(0,l2));

        final List<Integer> l3 = Arrays.asList(0, 1, 2, 3, 4, 5);
        assertSameLists(l3, uniqueArrayList);
    }


    @Test
    public void test_addAll_order_2() {
        UniqueArrayList<Integer> uniqueArrayList = new UniqueArrayList<>();

        final List<Integer> l1 = Arrays.asList(1, 2, 3, 4, 5);
        assertTrue(uniqueArrayList.addAll(l1));
        assertSameLists(l1, uniqueArrayList);

        final List<Integer> l2 = Arrays.asList(0, 1, 2, 3, 4, 5);
        assertTrue(uniqueArrayList.addAll(0,l2));

        assertSameLists(l2, uniqueArrayList);
    }
```

**Review by david-allison-1 — COMMENTED** [review]
```
/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:193: error: cannot find symbol
    private Optional<E> setAux(int index, E element) {
            ^
  symbol:   class Optional
  location: class UniqueArrayList<E>
  where E is a type-variable:
    E extends Object declared in class UniqueArrayList
1 error
```

**Comment by Arthur-Milchior** [comment]
> Issue is I don't know if from 1 year from now it will be misused, I can throw an exception when trying to get a ListIterator and do call the sorting methods internally instead of going through outside APIs for sorting

Thanks for caring about long term.
My point is that, if you data structure don't extends List/Set interface, then you can be sure that it won't be used with the Collection API, as the type will stop it. That seems safer

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java (hunk 12 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:256]: My bad. I thought that it was using "setAt" and not "addAt". I was indeed wrong here.

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java [hunk 13]
```diff
@@ -0,0 +1,419 @@
...
+     * @param c collection containing elements to be added to this collection
+     * @return true if this collection changed as a result of the call
+     */
+    @Override
+    public boolean addAll(final int index, @NonNull Collection<? extends E> c) {
+        checkIndex(index, true);
+        int cIndex = index;
+        boolean changed = false;
+        for (E e : c) {
+            boolean status = addAt(cIndex, e);
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:266]: That's not really efficient; if you add N elements in a list of size L, it'll take O(NL) time. If you don't use it that's not really a trouble.

Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:266]: Yeah, I noticed that, but I implemented to satisfy the List/Set interfaces

Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:266]: It will be used once when dialog startup, filling the tags found in checked tags but not in all tags.

**Comment by TarekkMA** [comment]
@Arthur-Milchior  the more I refactor `UniqueArrayList<E>` the more I find it similar to `SetUniqueList<E>`  🤦 

`SetUniqueList<E>` constructor accepts a generic list and set (with can be tree set). if I extend `SetUniqueList<E>` I would only write some factory methods to initialize the list and set add a sort method and that's it.

Thier implantation is the same but efficient 😄 

Docs:
https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/list/SetUniqueList.html
Source Code:
https://commons.apache.org/proper/commons-collections/jacoco/org.apache.commons.collections4.list/SetUniqueList.java.html


At least we would keep the tests :D

**Comment by TarekkMA** [comment]
Sorry if that PR wasted a lot of review time 😟 I should have done more extensive research. Work smart not hard I guess

File: AnkiDroid/src/test/java/com/ichi2/utils/UniqueArrayListTest.java [hunk 14]
```diff
@@ -235,6 +209,7 @@ public void test_set_not_existing() {
 
 
     @Test
+    @Ignore("SetUniqueList behaved different than the last implementation")
```
Comment (TarekkMA) [inline AnkiDroid/src/test/java/com/ichi2/utils/UniqueArrayListTest.java:212]: `SetUniqueList` does set first then remove duplicates, which is the reverse of the previous implementation.  
```java
    /**
     * Sets the value at the specified index avoiding duplicates.
     * <p>
     * The object is set into the specified index. Afterwards, any previous
     * duplicate is removed. If the object is not already in the list then a
     * normal set occurs. If it is present, then the old version is removed.
     *
     * @param index  the index to insert at
     * @param object  the object to set
     * @return the previous object
     */
    @Override
    public E set(final int index, final E object) {
        final int pos = indexOf(object);
        final E removed = super.set(index, object);

        if (pos != -1 && pos != index) {
            // the object is already in the unique list
            // (and it hasn't been swapped with itself)
            super.remove(pos); // remove the duplicate by index
        }

        set.remove(removed); // remove the item deleted by the set
        set.add(object); // add the new item to the unique set

        return removed; // return the item deleted by the set
    }
```

**Comment by Arthur-Milchior** [comment]
May I ask how you did find SetUniqueList ? I had never heard of it, I don't really know the apache library.
I did some quick search myself, didn't find anything relevant. I honestly don't know how you could have known more search was required, unless you created a anki deck of all classes in the various standard libraries.
No problem for the review. I don't know you yet, so it's also hard because I don't know how you'd take reviews and I asked a lot of change. Nice to see that you accepted the remarks and improved the code, and didn't decide to close or abandon the PR because of it. I can never really know in advance with new contributors I've not interacted with a lot

**Comment by Arthur-Milchior** [comment]
Can you please merge both commits, no reason to keep code just to delete it after that

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+        }
+
+        Set<E> set;
+        if (comparator == null) {
+            set = new HashSet<>();
+        } else {
+            set = new TreeSet<>(comparator);
+        }
+
+        if (list.isEmpty()) {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:112]: Why do you need this special case ? 
This means that if an empty list is passed, it is used by the UniqueArrayList, but if a non empty list is passed, then it's not used.
It means that if you do
```java
ArrayList l = new ArrayList();
UniqueArrayList u = UniqueArrayList.from(l, null);
l.add(5)
```
then you break the class assumption that the underlying list is not modified externally but 
```java
ArrayList l = new ArrayList();
l.add(5);
UniqueArrayList u = UniqueArrayList.from(l, null);
l.add(6)
```
would be acceptable, because in this case `l` is not inside anymore

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+        return sl;
+    }
+
+    public void sort() {
+        sort(null);
+    }
+
+    @Override
+    public void sort(@Nullable Comparator<? super E> c) {
+        Collections.sort(mList, c);
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:129]: Don't you want to call super.sort?

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+ * <p>
+ * The implementation of this class uses {@link ArrayList} and {@link TreeSet}.
+ * </p>
+ */
+public class UniqueArrayList<E> extends SetUniqueList<E> implements List<E>, Set<E> {
+
+    /**
+     * Internal list
+     *
+     * Only used for sorting, modifying this list may result in breaking the class
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:55]: This comment is not clear until I read a lot about the class. I'd write

> This is the same list as the one used in SetUniqueList. We keep track of it in order to be able to sort it, so that `SetUniqueList` get access to a sorted list.  `SetUniqueList` requires that the list passed to it be not modified. We can actually do it without breaking the real assumptions, which is that each elements of the list are distincts and corresponds to the content of the set used in SetUniqueList.
> Java 8 contains a List.sort function, however we can't use it with API 21 so we need to rewrite it.

Can you please confirm that this is the reason why you need to create an extra class ? as other wise I would have excepted you to just use List.sort

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+import java.util.Set;
+import java.util.Spliterator;
+import java.util.TreeSet;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.annotation.RequiresApi;
+
+/**
+ * A unique collection of items, fast random access, lookup, maintaining order, and sorting.
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:36]: This is essentially a `SetUniqueList`, the only difference being that its elements can be sorted in a way `SetUniqueList` usually can't

**Comment by TarekkMA** [comment]
I did some search and I saw the name on some answers on stack overflow, and I know that apache libraries provide tons and tons of useful common features that can make development in java easier, and I guess they are well tested. Infact AnkiDroid uses it's compression library `implementation 'org.apache.commons:commons-compress:1.12'`.

About changes/reviews/comments, they don't irritate me at all the more change you as the more I learn, I'm still learning how to be a part of an open source organization, so I guess more changes make me happy 😄 

I made 2 different commits so if the new implementation is not okay, we can go back to the old one.

One thing that needs review is the new library set behavior.  (I have made a comment on it), we can modify it or leave it.

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java (hunk 16 above)
Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:129]: No, the sorting of the super relies on `List.sort` which is API24 feature. and the `ListIterator` of the super (`SetUniqueList`) don't allow setting items it throws an exception.

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java (hunk 17 above)
Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:55]: I would modify that both `List.sort` and `Collection.sort` requires a `ListIterator` that allows setting items.

Issue is the super `SetUniqueList` wraps the underlying  `ListIterator` to disallows any set operations. I don't really know the reason behind this becase they can set old item with new item and remove the old from set, but that is the implementation they have.

<details>
<Summary> `SetUniqueList` `ListIterator` Implementation </summary>

```java
    @Override
    public ListIterator<E> listIterator() {
        return new SetListListIterator<>(super.listIterator(), set);
    }
```

```java
    /**
     * Inner class iterator.
     */
    static class SetListListIterator<E> extends
            AbstractListIteratorDecorator<E> {

        private final Set<E> set;
        private E last = null;

        protected SetListListIterator(final ListIterator<E> it, final Set<E> set) {
            super(it);
            this.set = set;
        }

        @Override
        public E next() {
            last = super.next();
            return last;
        }

        @Override
        public E previous() {
            last = super.previous();
            return last;
        }

        @Override
        public void remove() {
            super.remove();
            set.remove(last);
            last = null;
        }

        @Override
        public void add(final E object) {
            if (set.contains(object) == false) {
                super.add(object);
                set.add(object);
            }
        }

        @Override
        public void set(final E object) {
            throw new UnsupportedOperationException("ListIterator does not support set");
        }
    }
```


</details>

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java (hunk 18 above)
Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:36]: Yeah, this is an old comment will update it. Also, another modification allow the use of a comparator.

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java (hunk 15 above)
Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:112]: Yes I should remove that. I want to make a new copy of the passed list even if it is empty to not introduce bugs in the code later

**Comment by TarekkMA** [comment]
> I'm not against this per se, I just don't understand the point? I feel like I've walked into a data structures class, but we're a flashcard app. Why are we spending time creating and maintaining data structure implementations?

This class is to be used in `TagsDialog` or in anywhere we need to maintain a "list" of tags, in the `TagsDialog` for example the list should be unique and case-insensitive but  also need to have fast random access by index, so it can be used by a RecyclerView. And also sortable.

<hr>

https://github.com/ankidroid/Anki-Android/blob/3dd1e8506425add2552bf65ada1664b76eec321c/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/TagsDialog.java#L243-L263

Legacy implementation allowed tags of different cases to be entered, "Tag", "TAG", "tag" all of those will be a separate tag and can be checked and added to the database, which is different of how the upstream handles it.

Also, the legacy uses `list.contains` which will be slower for users with large collections.

But you are right, my previous attempt of implementing this my self was a waste of time and error-prone, but I think the last commit where I use `SetUniqueList` and adding despaired functionality is much better.

File: AnkiDroid/src/test/java/com/ichi2/utils/UniqueArrayListTest.java (hunk 14 above)
Comment (TarekkMA) [reply AnkiDroid/src/test/java/com/ichi2/utils/UniqueArrayListTest.java]: Adapted the test for the new behavior of `SetUniqueList`.

**Comment by TarekkMA** [comment]
@Arthur-Milchior Made the required changes: beefed up the docs, and added the suggested fixes

**Comment by Arthur-Milchior** [comment]
@mikehardy wrote:
> I'm not against this per se, I just don't understand the point?

As the initial message stated, it is part of a bigger PR #8322. This PR required a data structure that is not part of API 21, so it has to be added. I did ask @TarekkMA to split the PR so that I can concentrate first on the data structure, and once it is merged, I can concentrate on the actual feature, using this data structure. Considering both the data structure and the feature together is complex, as both requires very different review and comments

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+import androidx.annotation.Nullable;
+import androidx.annotation.RequiresApi;
+
+/**
+ * A collection of items that doesn't allow duplicate items, and allows fast random access, lookup, maintaining order, and sorting.
+ * <p>
+ * The <code>List</code> interface makes certain assumptions/requirements. This
+ * implementation breaks these in certain ways, but this is merely the result of
+ * rejecting duplicates. Each violation is explained in the method, but it
+ * should not affect you. Bear in mind that Sets require immutable objects to
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:41]: There seems to be no point about mentionning "set" here. Not before you indicates that it is both a List and a Set.

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+ * function correctly.
+ *
+ * @implNote The implementation of this class extends {@link SetUniqueList<E>} and adds the ability to define a comparator
+ * to be used to judge uniqueness of elements, and allows sorting.
+ * <p>
+ * Data structures used internally:
+ * - {@link ArrayList<E>} to enable fast random access, sorting, and maintaining order of items during iteration
+ * - {@link Set<E>} to enable fast lookup
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:49]: Set is NOT a data structure. At most it can be called an abstract data structure. The difference being that `set` do not correspond to a specific implementation while `HashSet` and `TreeSet` do.

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+        }
+
+        final UniqueArrayList<E> sl = new UniqueArrayList<>(new ArrayList<>(), set);
+
+        sl.addAll(source);
+
+        return sl;
+    }
+
+    public void sort() {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:131]: Please add javadoc on both sort to indicate that this is equivalent to Collection.sort, using [natural ordering](https://docs.oracle.com/javase/7/docs/api/java/lang/Comparable.html) or given comparator

**Review by Arthur-Milchior — APPROVED** [review]
I approve only comment change required now.

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java (hunk 19 above)
Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:41]: I did copy this form the super class docs as is.

File: AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java (hunk 20 above)
Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:49]: I meant to be abstract and give the exact DS used if the comparetor is used or not. But. Will try to rephrase it to be correct

**Comment by TarekkMA** [comment]
@Arthur-Milchior 
Added docs for both sort methods, adapted from the `Collections#sort` docs, and cleared the class docs

File: AnkiDroid/build.gradle [hunk 22]
```diff
@@ -262,6 +262,7 @@ dependencies {
     // io.github.java-diff-utils:java-diff-utils is the natural successor here, but requires API24, #7091
     implementation 'org.bitbucket.cowwoc:diff-match-patch:1.2'
     implementation 'org.apache.commons:commons-compress:1.12' // #6419 - handle >2GB apkg files
+    implementation group: 'org.apache.commons', name: 'commons-collections4', version: '4.4' // SetUniqueList
```
Comment (mikehardy) [inline AnkiDroid/build.gradle:265]: I'd prefer to follow the same style as all the other dependencies, no reason to have one be different

```suggestion
    implementation 'org.apache.commons:commons-collections4:4.4' // SetUniqueList
```

**Review by mikehardy — APPROVED** [review]
Looks good to me - you've got one clean commit here - I would like my suggestion on the gradle dependency style merged, but if you want to fix it locally and re-push that would make sense. Otherwise we can squash-merge and release it :rocket:

**Comment by TarekkMA** [comment]
Thank you, Mike. Now I will continue on #8322 as there are not more pending PRs.

**Comment by TarekkMA** [comment]
Added your suggestion, I think squash and rebase.

**Review by Arthur-Milchior — APPROVED** [review]
